### PR TITLE
Install Elastic

### LINKED ISSUE #423: Bundle Elastic with installer
https://github.com/Gold872/elastic-dashboard

### PR DESCRIPTION
**Pull request by Gold872**
Resolves #423

### CODE REVIEW

File: scripts/elastic.gradle [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+def fileNameWindows = 'Elastic-WPILib-Windows.zip'
+
+def downloadUrlWindows = baseUrl + fileNameWindows
+
+def fileNameMac = 'Elastic-WPILib-macOS.zip'
+
+def downloadUrlMac = baseUrl + fileNameMac
+
+def fineNameMacArm = 'Elastic-WPILib-macOS.zip'
```
Comment (Gold856) [inline scripts/elastic.gradle:13]:
```suggestion
def fileNameMacArm = 'Elastic-WPILib-macOS.zip'
```

**Review by Gold856 — CHANGES_REQUESTED** [review]

File: files/Elastic.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+OS_NAME="$(uname -s)"
+ELASTIC_PATH="$(realpath "$SCRIPT_PATH/../elastic")"
+
+if [ "$OS_NAME" = "Linux" ]; then
+    exec "$ELASTIC_PATH/elastic_dashboard"
+elif [ "$OS_NAME" = "Darwin" ]; then
+    open "$ELASTIC_PATH/elastic_dashboard.app"
+else
+    exec "$ELASTIC_PATH/elastic_dashboard"
+fi
```
Comment (spacey-sooty) [inline files/Elastic.sh:15]:
```suggestion
if [ "$OS_NAME" = "Darwin" ]; then
    open "$ELASTIC_PATH/elastic_dashboard.app"
else
    exec "$ELASTIC_PATH/elastic_dashboard"
fi
```

**Review by spacey-sooty — CHANGES_REQUESTED** [review]

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 3]
```diff
@@ -801,6 +802,7 @@ private async Task RunShortcutCreator(CancellationToken token)
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "tools", "DataLogTool.exe"), $"{frcYear} WPILib Tools/Data Log Tool {frcYear}", $"Data Log Tool {frcYear}", ""));
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "advantagescope", "AdvantageScope (WPILib).exe"), $"{frcYear} WPILib Tools/AdvantageScope (WPILib) {frcYear}", $"AdvantageScope (WPILib) {frcYear}", ""));
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "choreo", "choreo.exe"), $"{frcYear} WPILib Tools/Choreo {frcYear}", $"Choreo {frcYear}", ""));
+            shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"{frcYear} WPILib Tools/Elastic ${frcYear}", $"Elastic {frcYear}", ""));
```
Comment (spacey-sooty) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:805]:
```suggestion
            shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"{frcYear} WPILib Tools/Elastic ${frcYear}", $"Elastic (WPILib) {frcYear}", ""));

```

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 4]
```diff
@@ -813,6 +815,7 @@ private async Task RunShortcutCreator(CancellationToken token)
             shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "tools", "DataLogTool.exe"), $"Programs/{frcYear} WPILib Tools/Data Log Tool {frcYear}", $"Data Log Tool {frcYear}", ""));
             shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "advantagescope", "AdvantageScope (WPILib).exe"), $"Programs/{frcYear} WPILib Tools/AdvantageScope (WPILib) {frcYear}", $"AdvantageScope (WPILib) {frcYear}", ""));
             shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "choreo", "choreo.exe"), $"Programs/{frcYear} WPILib Tools/Choreo {frcYear}", $"Choreo {frcYear}", ""));
+            shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"Programs/{frcYear} WPILib Tools/Elastic ${frcYear}", $"Elastic {frcYear}", ""));
```
Comment (spacey-sooty) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:818]:
```suggestion
            shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"Programs/{frcYear} WPILib Tools/Elastic (WPILib) ${frcYear}", $"Elastic (WPILib) {frcYear}", ""));

```

**Review by spacey-sooty — CHANGES_REQUESTED** [review]

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 5]
```diff
@@ -801,6 +802,7 @@ private async Task RunShortcutCreator(CancellationToken token)
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "tools", "DataLogTool.exe"), $"{frcYear} WPILib Tools/Data Log Tool {frcYear}", $"Data Log Tool {frcYear}", ""));
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "advantagescope", "AdvantageScope (WPILib).exe"), $"{frcYear} WPILib Tools/AdvantageScope (WPILib) {frcYear}", $"AdvantageScope (WPILib) {frcYear}", ""));
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "choreo", "choreo.exe"), $"{frcYear} WPILib Tools/Choreo (WPILib) {frcYear}", $"Choreo (WPILib) {frcYear}", ""));
+            shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"{frcYear} WPILib Tools/Elastic ${frcYear}", $"Elastic (WPILib) {frcYear}", ""));
```
Comment (calcmogul) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:805]:
```suggestion
            shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"{frcYear} WPILib Tools/Elastic (WPILib) ${frcYear}", $"Elastic (WPILib) {frcYear}", ""));

```

**Review by calcmogul — APPROVED** [review]

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 6]
```diff
@@ -813,6 +816,7 @@ private async Task RunShortcutCreator(CancellationToken token)
             shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "tools", "DataLogTool.exe"), $"Programs/{frcYear} WPILib Tools/Data Log Tool {frcYear}", $"Data Log Tool {frcYear}", ""));
             shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "advantagescope", "AdvantageScope (WPILib).exe"), $"Programs/{frcYear} WPILib Tools/AdvantageScope (WPILib) {frcYear}", $"AdvantageScope (WPILib) {frcYear}", ""));
             shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "choreo", "choreo.exe"), $"Programs/{frcYear} WPILib Tools/Choreo (WPILib) {frcYear}", $"Choreo (WPILib) {frcYear}", ""));
+            shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"Programs/{frcYear} WPILib Tools/Elastic (WPILib) ${frcYear}", $"Elastic (WPILib) {frcYear}", ""));
```
Comment (Gold856) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:819]:
```suggestion
            shortcutData.StartMenuShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"Programs/{frcYear} WPILib Tools/Elastic (WPILib) {frcYear}", $"Elastic (WPILib) {frcYear}", ""));

```

**Review by Gold856 — CHANGES_REQUESTED** [review]

File: WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs [hunk 7]
```diff
@@ -801,6 +802,7 @@ private async Task RunShortcutCreator(CancellationToken token)
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "tools", "DataLogTool.exe"), $"{frcYear} WPILib Tools/Data Log Tool {frcYear}", $"Data Log Tool {frcYear}", ""));
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "advantagescope", "AdvantageScope (WPILib).exe"), $"{frcYear} WPILib Tools/AdvantageScope (WPILib) {frcYear}", $"AdvantageScope (WPILib) {frcYear}", ""));
             shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "choreo", "choreo.exe"), $"{frcYear} WPILib Tools/Choreo (WPILib) {frcYear}", $"Choreo (WPILib) {frcYear}", ""));
+            shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"{frcYear} WPILib Tools/Elastic (WPILib) ${frcYear}", $"Elastic (WPILib) {frcYear}", ""));
```
Comment (sciencewhiz) [inline WPILibInstaller-Avalonia/ViewModels/InstallPageViewModel.cs:805]:
```suggestion
            shortcutData.DesktopShortcuts.Add(new ShortcutInfo(Path.Join(frcHomePath, "elastic", "elastic_dashboard.exe"), $"{frcYear} WPILib Tools/Elastic (WPILib) {frcYear}", $"Elastic (WPILib) {frcYear}", ""));

```

**Comment by sciencewhiz** [comment]
Elastic fails to run on macArm, likely because it isn't notarized.

**Comment by sciencewhiz** [comment]
I might have been chasing something else. After the installer is run, elastic isn't unzipped. When I manually unzip it, it runs. While I was troubleshooting, I downloaded Elastic direct from GitHub and after unzipping both the WPILib and the regular versions, they didn't run for me.

**Comment by sciencewhiz** [comment]
To fix the extraction, it needs code in the tools updater similar to AdvantageScope

https://github.com/wpilibsuite/WPILibInstaller-Avalonia/blob/0c7f6270c45eb09070e388b09c396092487bc273/apps/ToolsUpdater/src/main/java/Program.java#L98-L117
https://github.com/wpilibsuite/WPILibInstaller-Avalonia/blob/0c7f6270c45eb09070e388b09c396092487bc273/apps/ToolsUpdater/src/main/java/Program.java#L132-L133

**Comment by Gold872** [comment]
I'm looking into implementing this extraction, and I noticed that the AdvantageScope file extension for macos is a `.tar.gz` but Elastic is a `.zip`

I'm not entirely familiar with development on MacOS, is the `tar.gz` format required for portable apps on mac? I don't have a way of testing this on my machine.

**Comment by sciencewhiz** [comment]
I thought that it might be a problem, but the zip extracts and runs on my Mac, so I don't think it's necessary to change to .tar.gz, although that would be a more standard for Mac and Linux.

**Comment by Gold872** [comment]
When extracting it on macos, did you have to extract it twice? I noticed that there was a zip file in the zip file

**Comment by Gold872** [comment]
I just did some playing around with the CI, turns out it was putting the zip file within another zip file when uploading. There isn't an easy way to get around it besides manually extracting it before manually uploading it to the release. I'll copy the code that AdvantageScope has for extracting and use it for this. The changes won't work until the next release which should be somewhat soon.

**Comment by sciencewhiz** [comment]
> When extracting it on macos, did you have to extract it twice?

I didn't look specifically. The gui extracted automatically, but may have handled the double zipping seamlessly. 

.tar.gz is standard for Mac and Linux.

**Comment by Gold856** [comment]
The double zip is actually needed to ensure the bundle contents stay unchanged. If you try to upload the raw .app bundle as a GitHub Actions artifact, it will actually change permissions (and possibly some other things), which will causes the integrity checks to fail (since the contents aren't exactly the same). AdvantageScope is packaged the same way for Mac, with a double archive, one of which gets stripped during the installer build process, and then the final extraction happening on the user's computer. Elastic could stay as a zip, as long as macOS has `unzip` or another tool to extract zip files.

**Comment by Gold872** [comment]
I see, that explains why the double zip is needed. I'll probably keep Elastic in the `tar.gz` format since it seems to have more options with extractions.

The second extraction will keep failing until the next release since it's trying to find a `tar.gz` file that doesn't exist yet

**Comment by sciencewhiz** [comment]
> The changes won't work until the next release which should be somewhat soon.

How soon? WPILib is planning another beta release this weekend, and it would be very nice to get Elastic in.

**Comment by sciencewhiz** [comment]
> This does raise the question though of how to sign this with the WPILib signature, since at some point during the build process it would have to be double extracted, and then zipped again. Not sure how this should be done without having it lose the signature.

Normal extraction and compression shouldn't mess with the signature. It's just the way that Gradle does it that does.

**Comment by Gold872** [comment]
@sciencewhiz the new update + build should fix the macos extraction, let me know if there are any issues with it besides notarization

**Comment by sciencewhiz** [comment]
It's still double compressed, a tar.gz is inside the zip, so it doesn't get successfully extracted in the ToolsUpdater

**Comment by Gold872** [comment]
> It's still double compressed, a tar.gz is inside the zip, so it doesn't get successfully extracted in the ToolsUpdater

iirc isn't it supposed to be double compressed and it gets extracted twice during the installation?

**Comment by Gold872** [comment]
Looking at how advantagescope is structured with its wpilib mac release, I think I see what I was doing wrong, this should hopefully fix it

**Comment by sciencewhiz** [comment]
It's no longer double compressed, but the tar.gz didn't get extracted

**Comment by Gold872** [comment]
That's odd, it should have the same logic as advantagescope where it gets extracted during installation

**Comment by sciencewhiz** [comment]
It does extract on a fresh install. I think that's a problem unrelated to this PR that I've seen before

**Comment by sciencewhiz** [comment]
Windows failure if fixed by #470

**Review by sciencewhiz — APPROVED** [review]

**Pull request merged**